libs/@local/hash-validation/src/data_type.rs

```diff
@@ -0,0 +1,279 @@
+use core::fmt;
+use std::{borrow::Borrow, fmt::Formatter};
+
+use error_stack::{bail, ensure, Report, ResultExt};
+use serde_json::Value as JsonValue;
+use thiserror::Error;
+use type_system::{url::VersionedUrl, DataType, DataTypeReference};
+
+use crate::{
+    error::{Actual, Expected},
+    OntologyTypeProvider, Schema, Validate,
+};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum JsonValueType {
+    Null,
+    Boolean,
+    Number,
+    Integer,
+    String,
+    Array,
+    Object,
+}
+
+impl fmt::Display for JsonValueType {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Null => fmt.write_str("null"),
+            Self::Boolean => fmt.write_str("boolean"),
+            Self::Number => fmt.write_str("number"),
+            Self::Integer => fmt.write_str("integer"),
+            Self::String => fmt.write_str("string"),
+            Self::Array => fmt.write_str("array"),
+            Self::Object => fmt.write_str("object"),
+        }
+    }
+}
+
+impl From<&JsonValue> for JsonValueType {
+    fn from(value: &JsonValue) -> Self {
+        match value {
+            JsonValue::Null => Self::Null,
+            JsonValue::Bool(_) => Self::Boolean,
+            JsonValue::Number(_) => Self::Number,
+            JsonValue::String(_) => Self::String,
+            JsonValue::Array(_) => Self::Array,
+            JsonValue::Object(_) => Self::Object,
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum DataTypeConstraint {
+    #[error("the provided value is not equal to the expected value")]
+    Const {
+        actual: JsonValue,
+        expected: JsonValue,
+    },
+    #[error("unknown constraint: `{key}`")]
+    Unknown { key: String },
+}
+
+#[derive(Debug, Error)]
+pub enum DataValidationError {
+    #[error("the validator was unable to read the data type: `{id}`")]
+    DataTypeRetrieval { id: VersionedUrl },
+    #[error(
+        "the value provided does not match the data type, expected `{expected}`, got `{actual}`"
+    )]
+    InvalidType {
+        actual: JsonValueType,
+        expected: JsonValueType,
+    },
+    #[error("a constraint was not fulfilled")]
+    ConstraintUnfulfilled,
+    #[error("the schema contains an unknown data type: `{schema}`")]
+    UnknownType { schema: String },
+}
+
+impl<P: Sync> Schema<JsonValue, P> for DataType {
+    type Error = DataValidationError;
+
+    async fn validate_value<'a>(
+        &'a self,
+        value: &'a JsonValue,
+        _provider: &'a P,
+    ) -> Result<(), Report<DataValidationError>> {
+        match self.json_type() {
```

Comment:
I am not a particular fan of us using a string here, but I know that is more an issue with the `type-system`, than with this library.

Reply:
I agree that this should not be a string. I [added a task](https://linear.app/hash/issue/BP-60/dont-use-string-for-the-json-type-in-datatype-in-type-system) for that.